compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2630,4 +2630,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             _ => (false, ty),
         }
     }
+
+    /// Record a pattern that's invalid under Rust 2024 match ergonomics, along with a problematic
+    /// span, so that the pattern migration lint can desugar it during THIR construction.
+    fn add_rust_2024_migration_desugared_pat(
+        &self,
+        pat_id: HirId,
+        subpat_span: Span,
+        cutoff_span: Span,
+        detailed_label: &str,
+    ) {
+        // Try to trim the span we're labeling to just the `&` or binding mode that's an issue.
+        // If the subpattern span is a macro call site, no trimming will be done.
+        let source_map = self.tcx.sess.source_map();
+        let cutoff_span = source_map
+            .span_extend_prev_while(cutoff_span, char::is_whitespace)
+            .unwrap_or(cutoff_span);
+        let trimmed_span = subpat_span.until(cutoff_span);
+
+        // Only provide a detailed label if the problematic subpattern isn't from an expansion.
+        // In the case that it's from a macro, we'll add a more detailed note in the emitter.
+        let desc = if subpat_span.from_expansion() {
+            "default binding mode is reset within expansion"
+        } else {
+            detailed_label
+        };
+
+        self.typeck_results
+            .borrow_mut()
+            .rust_2024_migration_desugared_pats_mut()
+            .entry(pat_id)
+            .or_default()
+            .push((trimmed_span, desc.to_owned()));
```

Comment:
I'm not totally certain if the span logic is resilient enough to ensure that a trimmed/modified span will always have the same edition as one of the input spans. In this case, `pat_span` may be from edition 2024 and `inner_span` may be from 2021.

To avoid worrying about this subtle change in behavior, could you record the `pat_span` or perhaps just its edition in this table too, so we don't implicitly rely on `trimmed_span`'s edition which may be totally random depending on the impl of `Span::until` and `SourceMap::span_extend_prev_while`.

Reply:
ah, yeah. good point. I've opted to re-add the syntax context from `pat.span` to the trimmed span, to avoid needing a separate span or boolean field (and thus a tuple or struct). using `Span::trim_end` would be a less roundabout way of doing this, since it just uses `SpanData::with_hi` to do the trimming, but that would rely on `Span::trim_end` keeping this behavior. we could also inline it to use `with_hi` directly, but that's less readable.